crux-core/src/crux/api/v2/Document.java

```diff
@@ -0,0 +1,74 @@
+package crux.api.v2;
+
+import clojure.lang.IPersistentMap;
+import clojure.lang.Keyword;
+import clojure.lang.PersistentArrayMap;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static crux.api.v2.Attribute.attr;
+import static crux.api.v2.CruxId.*;
+
+public class Document {
+    private final Map<Attribute, Object> document;
+
+    private Document(Map<Attribute, Object> attrs) {
+        this.document = attrs;
+    }
+
+    public Object get(Attribute attr) {
+        return document.get(attr);
+    }
+
+    public Object get(String attr) {
+        return document.get(Attribute.attr(attr));
+    }
+
+    public CruxId getId() {
+        return cruxId(document.get(Attribute.attr("crux.db/id")));
+    }
+
+    public static Document document(CruxId id) {
+        Map<Attribute, Object> initialDoc = Collections.singletonMap(Attribute.attr("crux.db/id"), id.toEdn());
+        return new Document(initialDoc);
+    }
+
+    static Document document(Map<Keyword, ?> map) {
+        Map<Attribute, Object> newDoc = new HashMap<>();
+        for(Keyword key : map.keySet()) {
+            newDoc.put(Attribute.attr(key), map.get(key));
+        }
+        return new Document(newDoc);
```

Comment:
There is no need to use `HashMap` for internal state of the `Document` class! You might change it to `IPersistentMap` right way (and have a short-cut to simply assign it, if `PeristentMap` is already given by the core).

This is exactly why you do incapsulation in the first place :-) Then all your methods become simpler and much more efficient.

Reply:
yep, agreed!